src/ebonite/core/objects/core.py

```diff
@@ -492,38 +561,101 @@ def delete_image(self, image: 'Image', meta_only: bool = False, cascade: bool =
 
     @_with_meta
     def save(self):
+        """Saves task to meta repository and pushes unsaved datasets to dataset repository"""
+        self.push_datasets()
         self._meta.save_task(self)
 
-    def add_dataset(self, name, dataset: Union[DatasetSource, Any], target: Any = None):
+    def _resolve_dataset(self, dataset: AnyDataset, name: str) -> str:
+        """Resolves existing dataset or creates a new one
+
+        :param dataset: dataset name, Dataset, DatasetSource or raw data object
+        :param name: name for dataset if it is new"""
+        if isinstance(dataset, str):
+            if dataset not in self.datasets:
+                raise ValueError(f'no dataset named {dataset} in task {self}')  # TODO maybe ohter error?
+            return dataset
+
+        self.add_dataset(name, dataset)
+        return name
+
+    def _resolve_metric(self, metric: AnyMetric, name: str):
+        """Resolves existing metric or creates a new one
+
+        :param metric: metric name, Metric or raw metric object
+        :param name: name for metric if it is new"""
+        if isinstance(metric, str):
+            if metric not in self.metrics:
+                raise ValueError(f'no metric named {metric} in task {self}')  # TODO maybe ohter error?
+            return metric
+        self.add_metric(name, metric)
+        return name
+
+    def add_evaluation(self, name: str,
+                       data: AnyDataset,
+                       target: AnyDataset,
+                       metrics: Union[AnyMetric, List[AnyMetric]]):
+        """Adds new evaluation set to this task
+
+        :param name: name of the evaluation set
+        :param data: input dataset for evaluation
+        :param target: ground truth for input data
+        :param metrics: one or more metrics to measure"""
+        if name in self.evaluation_sets:
+            raise ValueError(f'evalset {name} already in task {self}')
+        data = self._resolve_dataset(data, f'{name}_input')
+        target = self._resolve_dataset(target, f'{name}_output')
+        if not isinstance(metrics, list):
+            metrics = [metrics]
+        metrics = [self._resolve_metric(m, f'{name}_{i}') for i, m in enumerate(metrics)]
+        self.evaluation_sets[name] = EvaluationSet(data, target, metrics)
+
+    @_with_dataset
+    def add_dataset(self, name, dataset: Union[DatasetSource, AbstractDataset, Any]):
```

Comment:
What about dataset removal? Aren't there supposed to be methods for that?

Reply:
good one